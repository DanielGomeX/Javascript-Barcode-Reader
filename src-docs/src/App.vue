<template>
  <div id="app">
    <NavBar></NavBar>
    <Header />

    <transition name="slide-up" appear="appear" mode="out-in">
      <router-view />
    </transition>
  </div>
</template>
<script>
// @ is an alias to /src
import NavBar from '@/components/NavBar.vue'
import Header from '@/components/Header.vue'

export default {
  components: {
    NavBar,
    Header,
  },
}
</script>
<style lang="scss"></style>
