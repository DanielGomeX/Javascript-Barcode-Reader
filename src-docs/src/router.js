import Vue from 'vue'
import Router from 'vue-router'
import GettingStarted from './views/GettingStarted.vue'
import Home from './views/Home.vue'
import Samples from './views/Samples.vue'

Vue.use(Router)

export default new Router({
  routes: [
    {
      path: '/home',
      name: 'home',
      component: Home,
    },
    {
      path: '/getting-started',
      name: 'getting-started',
      // route level code-splitting
      // this generates a separate chunk (about.[hash].js) for this route
      // which is lazy-loaded when the route is visited.
      component: GettingStarted,
    },
    {
      path: '/samples',
      name: 'samples',
      // route level code-splitting
      // this generates a separate chunk (about.[hash].js) for this route
      // which is lazy-loaded when the route is visited.
      component: Samples,
    },
    {
      path: '*',
      redirect: '/home',
    },
  ],
})
