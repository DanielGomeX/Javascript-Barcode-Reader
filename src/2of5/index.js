const CHAR_SET = [
  'nnwwn',
  'wnnnw',
  'nwnnw',
  'wwnnn',
  'nnwnw',
  'wnwnn',
  'nwwnn',
  'nnnww',
  'wnnwn',
  'nwnwn',
]
const WEIGHTS = [1, 2, 4, 7, 0]

module.exports = (lines, type = 'interleaved ') => {
  let code = []
  const binary = []
  const sequenceBar = []
  const sequenceSpace = []

  // remove empty whitespaces
  lines.shift()
  lines.pop()

  const barThreshold = Math.ceil(
    lines.reduce((pre, item) => pre + item, 0) / lines.length
  )

  // extract start pair
  const startChar = lines
    .splice(0, 4)
    .map(line => (line > barThreshold ? 1 : 0))
    .join('')
  const endChar = lines
    .splice(lines.length - 4, 4)
    .map(line => (line > barThreshold ? 1 : 0))
    .join('')

  console.log(startChar, endChar)

  for (let i = 0; i < lines.length; i += 1) {
    let segment = lines[i]

    const isEven = i % 2 === 0
    if (isEven) {
      sequenceBar.push(segment > barThreshold ? 'w' : 'n')
    } else {
      sequenceSpace.push(segment > barThreshold ? 'w' : 'n')
    }
  }

  // for (let i = 0; i < sequenceBar.length; i += 5) {
  //   const segBar = sequenceBar.slice(i, i + 5).join('')
  //   const segSpace = sequenceSpace.slice(i, i + 5).join('')

  //   console.log(segBar, segSpace)
  // }

  console.log(sequenceBar, sequenceSpace)

  return code.join('')
}
