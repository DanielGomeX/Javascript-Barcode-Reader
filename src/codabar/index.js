module.exports = lines => {
  // TODO: implement this
  console.log(lines)
}
