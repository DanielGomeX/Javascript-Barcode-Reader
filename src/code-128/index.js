const CHAR_SET = {}

module.exports = lines => {
  let code = []
  const binary = []
  const barWidths = []
  let linestemp = lines.slice()

  // remove termination bars
  // lines.pop()

  const barThreshold = lines.reduce((pre, item) => (pre + item) / 2, 0)

  let minBarWidth = lines.reduce((pre, item) => {
    if (item < barThreshold) return (pre + item) / 2
    return pre
  }, 0)
  // minBarWidth = mode(lines)
  // leave the padded *
  for (let i = 0; i < lines.length; i += 1) {
    let segment = lines[i]

    while (segment > 0) {
      if (i % 2 === 0) {
        binary.push(1)
      } else {
        binary.push(0)
      }
      segment -= minBarWidth
    }
  }
  linestemp.pop()
  while (linestemp.length > 0) {
    let seg = linestemp.splice(0, 6)
    let minWidth = seg.reduce((pre, item) => pre + item, 0) / 11
    barWidths.push(seg.map(item => Math.round(item / minWidth)))
  }
  // Start char
  const startChar = binary.splice(0, 11).join('')
  while (binary.length > 0) {
    let a = binary.splice(0, 11).join('')
  }
  let table = CTable

  switch (startChar) {
    case '11010000100':
      table = ATable
      break
    case '11010010000':
      table = BTable
      break
  }

  return code.join('')
}

function mode(array) {
  if (array.length == 0) return null
  let modeMap = {}
  let maxEl = array[0],
    maxCount = 1
  for (let i = 0; i < array.length; i++) {
    let el = array[i]
    if (modeMap[el] == null) modeMap[el] = 1
    else modeMap[el]++
    if (modeMap[el] > maxCount) {
      maxEl = el
      maxCount = modeMap[el]
    }
  }
  return maxEl
}
