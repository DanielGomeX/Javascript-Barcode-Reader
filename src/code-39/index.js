const BAR_SET = {
  '10001': '1',
  '01001': '2',
  '11000': '3',
  '00101': '4',
  '10100': '5',
  '01100': '6',
  '00011': '7',
  '10010': '8',
  '01010': '9',
  '00110': '10',
}

const GROUP_SET = {
  '01000': '0',
  '00100': '10',
  '00010': '20',
  '10000': '30',
}

const CHAR_SET = [
  '1',
  '2',
  '3',
  '4',
  '5',
  '6',
  '7',
  '8',
  '9',
  '0',
  'A',
  'B',
  'C',
  'D',
  'E',
  'F',
  'G',
  'H',
  'I',
  'J',
  'K',
  'L',
  'M',
  'N',
  'O',
  'P',
  'Q',
  'R',
  'S',
  'T',
  'U',
  'V',
  'W',
  'X',
  'Y',
  'Z',
  '-',
  '.',
  '␣',
  '*',
]

const decode = lines => {
  // manualy push last white space
  lines.push(3)
  let code = ''

  for (let i = 1; i < lines.length; i += 10) {
    const segment = lines.slice(i, i + 10)

    const barThreshold = Math.round(
      segment.reduce((pre, item) => pre + item, 0) / segment.length
    )

    const noob = segment.map(item => (item > barThreshold ? 1 : 0))
    const barSeg = noob.filter((item, index) => index % 2 === 0).join('')
    const whiteSeg = noob.filter((item, index) => index % 2 !== 0).join('')

    code +=
      CHAR_SET[
        parseInt(BAR_SET[barSeg], 10) - 1 + parseInt(GROUP_SET[whiteSeg], 10)
      ]
  }

  return code.substring(1, code.length - 1)
}

module.exports = {
  decode,
}
