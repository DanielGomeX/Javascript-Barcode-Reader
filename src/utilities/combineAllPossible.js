export async function combineAllPossible(finalResult, result) {
  let finalResultArr = finalResult.split('')
  let resultArr = result.split('')

  resultArr.forEach((char, index) => {
    if (!finalResultArr[index] || finalResultArr[index] === '?') {
      if (char && char !== '?') {
        finalResultArr[index] = char
      }
    }
  })

  return finalResultArr.join('')
}
