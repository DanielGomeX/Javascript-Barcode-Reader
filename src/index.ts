import { combineAllPossible } from './utilities/combineAllPossible'
import { getImageDataFromSource } from './utilities/getImageDataFromSource'
import { getLines } from './utilities/getLines'
import { applyAdaptiveThreshold } from './utilities/threshold/adaptiveThreshold'
import { applySimpleThreshold } from './utilities/threshold/applySimpleThreshold'

const BARCODE_DECODERS: {
  [key: string]: Function
} = {
  'code-128': require('./code-128'),
  'code-2of5': require('./2of5'),
  'code-39': require('./code-39'),
  'code-93': require('./code-93'),
  'ean-13': require('./ean-13'),
  'ean-8': require('./ean-8'),
  codabar: require('./codabar')
}

export async function javascriptBarcodeReader(
  image: string | HTMLImageElement | HTMLCanvasElement | ImageData,
  options: { barcode: string; type?: any; useAdaptiveThreshold?: any; singlePass?: any }
): Promise<string> {
  // store intermediary results, get final result by replacing ? from available result
  let finalResult = ''
  const barcode = options.barcode.toLowerCase()
  const type = (options.type || '').toLowerCase()

  const list = Object.keys(BARCODE_DECODERS)

  if (!list.includes(barcode)) {
    throw new Error(
      `Invalid barcode specified. Available decoders: ${list}. https://github.com/mubaidr/Javascript-Barcode-Reader#available-decoders`
    )
  }

  const imageData = await getImageDataFromSource(image)
  let data = imageData.data
  const width = imageData.width
  const height = imageData.height

  const channels = data.length / (width * height)

  // apply adaptive threshold
  if (options.useAdaptiveThreshold) {
    data = applyAdaptiveThreshold(data, width, height)
  }

  // check points for barcode location
  const sPoints = [5, 6, 4, 7, 3, 8, 2, 9, 1]
  const slineStep = Math.round(height / sPoints.length)
  //should be odd number to be able to find center
  const rowsToScan = Math.min(3, height)

  for (let i = 0; i < sPoints.length; i += 1) {
    const sPoint = sPoints[i]
    const start = channels * width * Math.floor(slineStep * sPoint)
    const end = start + rowsToScan * channels * width
    let dataSlice = data.slice(start, end)

    if (!options.useAdaptiveThreshold) {
      dataSlice = applySimpleThreshold(dataSlice, width, rowsToScan)
    }

    const lines = getLines(dataSlice, width, rowsToScan)

    if (!lines || lines.length === 0) {
      if (options.singlePass) throw new Error('Failed to extract barcode!')

      continue
    }

    // Run the decoder
    const result = BARCODE_DECODERS[barcode](lines, type)

    if (!result) continue
    if (result.indexOf('?') === -1) return result
    if (options.singlePass) return result

    if (finalResult === '') {
      finalResult = result
    } else {
      finalResult = combineAllPossible(finalResult, result)

      if (!finalResult.includes('?')) return finalResult
    }

    if (i === sPoints.length - 1) return finalResult
  }

  throw new Error('Failed to extract barcode!')
}
