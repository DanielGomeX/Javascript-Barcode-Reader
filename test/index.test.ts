import * as Jimp from 'jimp'
import * as path from 'path'
import { BARCODE_DECODERS, javascriptBarcodeReader } from '../src/index'
import { combineAllPossible } from '../src/utilities/combineAllPossible'

beforeAll(() => {
  jest.setTimeout(5000)
})

describe('combineAllPossible', () => {
  test('should be able to combine multiple results into one complete', () => {
    const result = combineAllPossible('?123456', '012345?')

    expect(result).toBe('0123456')
  })
})

describe('extract barcode from local files', () => {
  test('should detect barcode codabar', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/codabar.jpg'),
      barcode: BARCODE_DECODERS.codabar
    })

    expect(result).toBe('A40156C')
  })

  test('should detect barcode codabar', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/codabar.jpg'),
      barcode: BARCODE_DECODERS.codabar,
      options: {
        singlePass: true
      }
    })

    expect(result).toBe('A40156C')
  })

  test('should detect barcode 2 of 5 standard', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-2of5.jpg'),
      barcode: 'code-2of5'
    })

    expect(result).toBe('12345670')
  })

  test('should detect barcode 2 of 5 interleaved', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-2of5-interleaved.jpg'),
      barcode: 'code-2of5',
      barcodeType: 'interleaved'
    })

    expect(result).toBe('12345670')
  })

  test('should detect barcode 39', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-39.jpg'),
      barcode: 'code-39'
    })

    expect(result).toBe('10023')
  })

  test('should detect barcode 93', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-93.jpg'),
      barcode: 'code-93'
    })

    expect(result).toBe('123ABC')
  })

  test('should detect barcode 128', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-128.jpg'),
      barcode: 'code-128'
    })

    expect(result).toBe('ABC-abc-1234')
  })

  test('should detect barcode EAN-8', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/ean-8.jpg'),
      barcode: 'ean-8'
    })

    expect(result).toBe('73127727')
  })

  test('should detect barcode EAN-13', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/ean-13.jpg'),
      barcode: 'ean-13'
    })

    expect(result).toBe('901234123457')
  })

  test('should detect barcode 128 without padding white bars', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-128-no-padding.jpg'),
      barcode: 'code-128'
    })

    expect(result).toBe('12ab#!')
  })

  test('should detect barcode 128 with multiple zeros', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-128-000.jpg'),
      barcode: 'code-128'
    })

    expect(result).toBe('79619647103200000134407005')
  })

  test('should detect barcode 128 with default start Code B', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/L89HE1806005080432.gif'),
      barcode: 'code-128'
    })

    expect(result).toBe('L89HE1806005080432')
  })

  test('should detect barcode 93 without padding white bars', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-93-no-padding.jpg'),
      barcode: 'code-93'
    })

    expect(result).toBe('WIKIPEDIA')
  })

  test('should detect barcode 93 with bitmap data', async () => {
    const image = await Jimp.read('./test/sample-images/code-93-no-padding.jpg')
    const { data, width, height } = image.bitmap

    const result = await javascriptBarcodeReader({
      image: {
        data: Uint8ClampedArray.from(data),
        width,
        height
      },
      barcode: 'code-93'
    })

    expect(result).toBe('WIKIPEDIA')
  })
})

describe('extract barcode after applying adaptive threhsold', () => {
  test('should detect barcode codabar', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/codabar.jpg'),
      barcode: 'codabar',
      options: {
        useAdaptiveThreshold: true
      }
    })

    expect(result).toBe('A40156C')
  })

  test('should detect barcode 2 of 5', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-2of5.jpg'),
      barcode: 'code-2of5',
      options: {
        useAdaptiveThreshold: true
      }
    })

    expect(result).toBe('12345670')
  })

  test('should detect barcode 2 of 5 interleaved', async () => {
    const result = await javascriptBarcodeReader({
      image: path.resolve('./test/sample-images/code-2of5-interleaved.jpg'),
      barcode: 'code-2of5',
      barcodeType: 'interleaved',
      options: {
        useAdaptiveThreshold: true
      }
    })

    expect(result).toBe('12345670')
  })
})

describe('extract barcode from remote URL', () => {
  test('should detect barcode 93 from remote url', async () => {
    const result = await javascriptBarcodeReader({
      image: 'https://upload.wikimedia.org/wikipedia/en/a/a9/Code_93_wikipedia.png',
      barcode: 'code-93'
    })
    expect(result).toBe('WIKIPEDIA')
  })
})

describe('Fails', () => {
  test('throws when no barcode specified', async () => {
    try {
      await javascriptBarcodeReader({
        image: 'https://upload.wikimedia.org/wikipedia/en/a/a9/Code_93_wikipedia.png',
        barcode: 'oallal'
      })
    } catch (err) {
      expect(err).toBeDefined()
    }
  })

  test('throws when invalid barcode specified', async () => {
    try {
      await javascriptBarcodeReader({
        image: './test/sample-images/empty.jpg',
        barcode: 'none'
      })
    } catch (err) {
      expect(err).toBeDefined()
    }
  })

  test('throws when no barcode found', async () => {
    try {
      await javascriptBarcodeReader({
        image: './test/sample-images/empty.jpg',
        barcode: 'code-93'
      })
    } catch (err) {
      expect(err).toBeDefined()
    }
  })
})
