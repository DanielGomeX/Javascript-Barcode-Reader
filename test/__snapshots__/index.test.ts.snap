// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`isUrl check if string is URL 1`] = `
Uint8ClampedArray [
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
  0,
]
`;
